### PR TITLE
Added e2e test.

### PR DESCRIPTION
**Pull request by balajismaniam**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -29,13 +29,46 @@ jobs:
...
+    docker:
+      - image: volumecontroller/golang:1.9.2
+    working_directory: /go/src/github.com/NervanaSystems/kube-volume-controller
+    steps:
+      - checkout
+      - run:
+          name: install chart
+          command: |
+            mkdir -p $HOME/.kube
+            KUBE_CONFIG=${GKE_KUBECONFIG_KVCCI}
```
Comment (jose5918) [inline .circleci/config.yml]: @balajismaniam what cluster is this?

File: pkg/handlers/nfs_handler.go [hunk 2]
```diff
@@ -127,6 +127,10 @@ func (h *nfsHandler) OnDelete(ns string, vc kvcv1.VolumeConfig, controllerRef me
 
 		for _, resource := range resourceList {
 			resControllerRef := metav1.GetControllerOf(resource)
+			if resControllerRef == nil {
```
Comment (Ajay191191) [inline pkg/handlers/nfs_handler.go]: Does this mean we should just delete it?

Comment (Ajay191191) [reply pkg/handlers/nfs_handler.go]: Or maybe we should handle this in the reconciler. Can you add a TODO there.

File: pkg/handlers/s3_dev_handler.go [hunk 3]
```diff
@@ -165,22 +165,20 @@ func (h *s3DevHandler) OnAdd(ns string, vc kvcv1.VolumeConfig, controllerRef met
 }
 
 func (h *s3DevHandler) OnDelete(ns string, vc kvcv1.VolumeConfig, controllerRef metav1.OwnerReference) {
-	for _, client := range h.k8sResourceClients {
-		if client.Plural() == "nodes" {
-			continue
-		}
+	podClient := getK8SResourceClientFromPlural(h.k8sResourceClients, "pods")
```
Comment (Ajay191191) [inline pkg/handlers/s3_dev_handler.go]: We also need to delete the actual data on the node. That can go in a follow-up though.

File: pkg/handlers/s3_handler.go [hunk 4]
```diff
@@ -191,9 +191,12 @@ func (h *s3Handler) OnDelete(ns string, vc kvcv1.VolumeConfig, controllerRef met
 
 	for _, resource := range podList {
 		resControllerRef := metav1.GetControllerOf(resource)
+		if resControllerRef == nil {
```
Comment (Ajay191191) [inline pkg/handlers/s3_handler.go]: Same comment here as above. Maybe a TODO would help.

File: test/e2e/kvc_test.go [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+		// Positive test cases.
+		{
+			description: "single vc - S3 - no error",
+			volumeConfigs: []crv1.VolumeConfig{
+				{
+					ID:          "vol1",
+					Replicas:    1,
+					SourceType:  "S3",
+					SourceURL:   "s3://e2e-test/cifar-100-python.tar.gz",
+					EndpointURL: "http://10.63.255.0:9000",
```
Comment (Ajay191191) [inline test/e2e/kvc_test.go]: If the service restarts/changes, we might have to change the IP here. I think we should convert this to an argument too.

File: test/e2e/kvc_test.go [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+					ID:         "vol1",
+					SourceType: "NFS",
+					AccessMode: "ReadWriteMany",
+					Capacity:   ".5Gi",
+					Labels: map[string]string{
+						"key1": "val1",
+						"key2": "val2",
+					},
+					Options: map[string]string{
+						"server": "10.63.249.182",
```
Comment (Ajay191191) [inline test/e2e/kvc_test.go]: This too.

File: test/e2e/kvc_test.go [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+		},
+		{
+			description: "multiple vc - S3 and NFS - no error",
+			volumeConfigs: []crv1.VolumeConfig{
+				{
+					ID:          "vol1",
+					Replicas:    1,
+					SourceType:  "S3",
+					SourceURL:   "s3://e2e-test/cifar-100-python.tar.gz",
+					EndpointURL: "http://10.63.255.0:9000",
```
Comment (Ajay191191) [inline test/e2e/kvc_test.go]: Same here ...

File: test/e2e/kvc_test.go [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+					ID:         "vol1",
+					SourceType: "NFS",
+					AccessMode: "ReadWriteMany",
+					Capacity:   ".5Gi",
+					Labels: map[string]string{
+						"key3": "val3",
+						"key4": "val4",
+					},
+					Options: map[string]string{
+						"server": "10.63.249.182",
```
Comment (Ajay191191) [inline test/e2e/kvc_test.go]: Same here ...

File: test/e2e/kvc_test.go [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+					},
+				},
+			},
+			expSuccess: false,
+			expError:   fmt.Sprintf("path has to be set in options"),
+			expHP:      false,
+			expNA:      false,
+			expPVC:     false,
+		},
```
Comment (Ajay191191) [inline test/e2e/kvc_test.go]: I think these negative tests are also tested in the unit tests. The one's we should negative test here are the cluster level things like:
1. minio endpoint is not reachable
2. s3 path is incorrect
3. nfs server ip is unreachable
4. nfs path is not correct.
5. One volumeConfig passes while the other one fails.

Maybe they can go in a follow-up PR too and we can just add a TODO here.

File: test/e2e/kvc_test.go [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+	}
+
+	for _, testCase := range testCases {
+		volman := makeVolumeManager(testCase.volumeConfigs)
+		createdVolman, err := crdClient.Create(volman)
+		require.Nil(t, err)
+		/*
```
Comment (Ajay191191) [inline test/e2e/kvc_test.go]: We can uncomment this before merging it in.

File: test/e2e/kvc_test.go [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+			expSuccess: false,
+			expError:   fmt.Sprintf("path has to be set in options"),
+			expHP:      false,
+			expNA:      false,
+			expPVC:     false,
+		},
+	}
+
+	for _, testCase := range testCases {
+		volman := makeVolumeManager(testCase.volumeConfigs)
```
Comment (Ajay191191) [inline test/e2e/kvc_test.go]: Just as a curiosity, why is it `volman` here and everywhere below?

File: test/e2e/kvc_test.go [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+		createdVolman, err := crdClient.Create(volman)
+		require.Nil(t, err)
+		/*
+			defer func() {
+				delOpts := &metav1.DeleteOptions{}
+				crdClient.Delete(volman.GetName(), delOpts)
+			}()
+		*/
+
+		if testCase.expSuccess {
```
Comment (Ajay191191) [inline test/e2e/kvc_test.go]: Here I think we should check if there exists a hostpath/NA/PVC for a particular id. Like we should check if hostpath and NA are created for s3 source type, and PVC is created for nfs type. We should add a TODO here too and add it in as a enhancement to the e2e tests.

Comment (Ajay191191) [reply test/e2e/kvc_test.go]: Also we should check for PVs along with PVCs.

File: pkg/handlers/nfs_handler.go (hunk 2 above)
Comment (balajismaniam) [reply pkg/handlers/nfs_handler.go]: No we should not delete these resources. They should not manipulated. This case is true when a resource has the same labels as a volume manager CR.

File: pkg/handlers/s3_dev_handler.go (hunk 3 above)
Comment (balajismaniam) [reply pkg/handlers/s3_dev_handler.go]: Agreed. But let's add it in a follow-up PR with the logic for garbage collection.

File: pkg/handlers/s3_handler.go (hunk 4 above)
Comment (balajismaniam) [reply pkg/handlers/s3_handler.go]: No we should not delete these resources. They should not manipulated. This case is true when a resource has the same labels as a volume manager CR.

File: test/e2e/kvc_test.go (hunk 9 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: Ok. I will work on adding these cases.

File: test/e2e/kvc_test.go (hunk 11 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: `volman` is short for Volume Manager.

File: test/e2e/kvc_test.go (hunk 5 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: I converted this to an arg.

File: test/e2e/kvc_test.go (hunk 6 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: I converted this to an arg.

File: test/e2e/kvc_test.go (hunk 7 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: I converted this to an arg.

File: test/e2e/kvc_test.go (hunk 8 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: I converted this to an arg.

File: test/e2e/kvc_test.go (hunk 9 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: I've added all the above test cases.

File: test/e2e/kvc_test.go (hunk 10 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: Done.

File: test/e2e/kvc_test.go (hunk 12 above)
Comment (balajismaniam) [reply test/e2e/kvc_test.go]: I added a TODO for this. Will create issues.

**Pull request merged**